contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,377 @@
+
+The snapshotter creates CreateSnapshotRequest and calls CreateSnapshot through CSI ControllerClient. It gets CreateSnapshotResponse from the out-of-tree CSI Volume Driver and creates a VolumeSnapshotContent API object with VolumeSnapshotSource.
+
+The snapshotter checks status in VolumeSnapshot to decide whether to bind. It binds VolumeSnapshot and VolumeSnapshotContent when ready.
+
+When the storage system fails to create snapshot, retry will not be performed in the first version. This is because users may not want to retry when taking consistent snapshots or scheduled snapshots when the timing of the snapshot creation is important. In a future version, a maxRetries flag will be added to allow users to control whether retries are needed.
+
+
+### Changes in PV Controller and CSI External Provisioner
+
+Both the CSI External Provisioner and the in-tree PV Controller will be modified to support provisioning a volume from a snapshot data source.
+
+`DataSource` is added in both `StorageClass` and `PersistentVolumeClaim` to represent the source of the data which is prepopulated to the provisioned volume. If `DataSource` is added to both during volume provisioning, `DataSource` in `PersistentVolumeClaim` will override `DataSource` in `StorageClass`. 
+
+The operation of the provisioning of a volume from a snapshot data source will be handled by the out-of-tree CSI External Provisioner.
+
+The in-tree PV Controller will handle the binding of the PV and PVC once they are ready.
+ 
+
+### CSI Volume Driver Snapshot Support
+
+The out-of-tree CSI Volume Driver creates a snapshot on the backend storage system or cloud provider, and calls CreateSnapshot through CSI ControllerServer and returns CreateSnapshotResponse. The out-of-tree CSI Volume Driver needs to implement the following functions:
+
+* CreateSnapshot, DeleteSnapshot, and create volume from snapshot if it supports CREATE_DELETE_SNAPSHOT.
+* ListSnapshots if it supports LIST_SNAPSHOTS.
+
+ListSnapshots can be an expensive operation because it will try to list all snapshots on the storage system. For a storage system that takes nightly periodic snapshots, the total number of snapshots on the system can be huge. Kubernetes should try to avoid this call if possible. Instead, calling ListSnapshots with a specific snapshot_id as filtering to query the status of the snapshot will be more desirable and efficient.
+
+CreateSnapshot is a synchronous function and it must be blocking until the snapshot is cut. For cloud providers that support the uploading of a snapshot as part of the creating snapshot operation, CreateSnapshot function must also be blocking until the snapshot is cut and after that it shall return an operation pending gRPC error code until the uploading process is complete.
+
```

Comment:
You mention that is is a "synchronous function" but then state that operation pending will be returned until the upload is complete. 
Does this mean that while a snapshot is in progress on a volume you can't take another snapshot until the current one has finished? (In some cases uploadin to cloud?)
This seems overly restrictive.

Reply:
"Synchronous" refers to the CSI driver interface part. Actually all the volume operations such as create snapshot, create volume are synchronous in CSI.  You can still take another snapshot while the current one is in progress.  If you call create snapshot with the same input parameters while the current one is in progress (pending), the call will be pending as well because it is idempotent.